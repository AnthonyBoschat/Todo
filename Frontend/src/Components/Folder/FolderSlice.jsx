import {createSlice} from "@reduxjs/toolkit"

const todoStorage = JSON.parse(localStorage.getItem("todoStorage"))

const FolderSlice = createSlice({
    name:"folder",
    initialState:{
        allFoldersLoad:false,
        folderOnCreation:false,
        folderSelectedID:null,
        folderSelectedName:null,
        foldersList:null
    },
    reducers:{
        update_allFoldersLoad:(state,action) => {
            state.allFoldersLoad = action.payload
        },
        update_folderOnCreation:(state,action) => {
            state.folderOnCreation = action.payload
        },
        update_folderSelectedID:(state,action) => {
            state.folderSelectedID = action.payload
        },
        update_folderSelectedName:(state,action) => {
            state.folderSelectedName = action.payload
        },
        update_loadFoldersList:(state,action) => {
            state.foldersList = action.payload
        },
        update_folderRename:(state,action) => {
            state.foldersList[action.payload.folderIndex].name = action.payload.newFolderName
        },
        update_addFolder:(state,action) => {
            state.foldersList.push(action.payload)
        },
        update_deleteFolder:(state,action) => {
            state.foldersList.splice(action.payload, 1)
        },
    },
})

export const FolderSliceReducer = FolderSlice.reducer
export const {
    update_allFoldersLoad,
    update_folderOnCreation,
    update_folderSelectedID,
    update_folderSelectedName,
    update_folderRename,
    update_loadFoldersList,
    update_addFolder,
    update_deleteFolder,
} = FolderSlice.actions