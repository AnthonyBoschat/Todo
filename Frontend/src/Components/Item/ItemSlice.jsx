import {createSlice} from "@reduxjs/toolkit"

const ItemSlice = createSlice({
    name:"item",
    initialState:{
        global:{
            itemToShow:[],
            itemOnCreation:false,
            
            tabs:{
                tabsList:["Items", "Property"],
                tabSelected:"Items"
            }
        }
        
    },
    reducers:{
        update_ItemOnCreation:(state,action) => {
            state.global.itemOnCreation = action.payload
        },
        update_itemToShow:(state,action) => {
            state.global.itemToShow = action.payload
        },
        update_addItemToShow:(state,action) => {
            state.global.itemToShow.push(action.payload)
        },
        update_tabSelectedItem:(state,action) => {
            state.global.tabs.tabSelected = action.payload
        }
    },
})

export const ItemSliceReducer = ItemSlice.reducer
export const {
    update_ItemOnCreation,
    update_itemToShow,
    update_addItemToShow,
    update_itemSelectedID,
    update_tabSelectedItem
} = ItemSlice.actions