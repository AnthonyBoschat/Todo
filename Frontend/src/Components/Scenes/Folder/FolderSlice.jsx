import {createSlice} from "@reduxjs/toolkit"

const todoStorage = JSON.parse(localStorage.getItem("todoStorage"))

const FolderSlice = createSlice({
    name:"folder",
    initialState:{
        folderOnCreation:false,
        folderSelectedID:null,
        folderSelectedName:null,
    },
    reducers:{
        update_folderOnCreation:(state,action) => {
            state.folderOnCreation = action.payload
        },
        update_folderSelectedID:(state,action) => {
            state.folderSelectedID = action.payload
        },
        update_folderSelectedName:(state,action) => {
            state.folderSelectedName = action.payload
        },
    },
})

export const FolderSliceReducer = FolderSlice.reducer
export const {
    update_folderOnCreation,
    update_folderSelectedID,
    update_folderSelectedName
} = FolderSlice.actions