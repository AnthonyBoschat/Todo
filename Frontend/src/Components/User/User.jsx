import React from "react";
import { useSelector } from "react-redux";
import useUser from "./useUser";

export default function User(){

    const userName = useSelector(store => store.connection.connectedUser.name)
    const {
        handleClickDisconnection
    } = useUser()
    return(
        <>
            {userName && (
                <div className="user_Display">
                    <div className="user_Box">
                        {userName}
                    </div>
                    <i onClick={handleClickDisconnection} className="disconnection fa-solid fa-arrow-right-from-bracket"></i>
                </div>
            )}
        </>
    )
}