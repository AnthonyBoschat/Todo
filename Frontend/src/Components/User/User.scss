div.user_Display{
    // background-color: rgba(140, 255, 0, 0.237); //
    display: flex;
    justify-content: center;
    align-items: center;
    gap: 0.4rem;

    div.user_Box{
        margin: 0.5rem;
        color: white;
        // outline: 1px solid white; //
        font-size: 1.5rem;
        height: fit-content;

        &:hover{
            cursor: default;
        }
        
    }
    i.disconnection{
        position: relative;
        top: 2px;
        align-self: center;
        color: white;
        opacity: 0.5;
        transition: opacity 0.05 linear;

        &:hover{
            opacity: 1;
            cursor: pointer;
        }
    }
}