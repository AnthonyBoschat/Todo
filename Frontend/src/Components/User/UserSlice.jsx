import {createSlice} from "@reduxjs/toolkit"

const UserSlice = createSlice({
    name:"user",
    initialState:{
        datas:{
            userFoldersList:[],
            userItemsList:[],
            userListsList:[],
            userPropertiesList:[]
        },
        allDatasLoad:false
    },
    reducers:{
        update_loadAllDatas:(state,action) => {
            const newDatas = {
                userFoldersList:action.payload.newUserFoldersList,
                userItemsList:action.payload.newUserItemsList,
                userListsList:action.payload.newUserListsList,
                userPropertiesList:action.payload.newUserPropertiesList
            }
            state.datas = newDatas
        },
        update_dataList:(state,action) => {
            state.datas[action.payload.listName] = action.payload.newList
        },
        update_addData:(state,action) => {
            state.datas[action.payload.listName].push(action.payload.newData)
        },
        update_deleteData:(state,action) => {
            state.datas[action.payload.listName].splice(action.payload.dataIndex, 1)
        },
        update_changeData:(state,action) => {
            state.datas[action.payload.listName][action.payload.dataIndex] = action.payload.newData
        },
        update_allDatasLoad:(state,action) => {
            state.allDatasLoad = action.payload
        },
        update_reorderList:(state,action) => {
            state.datas[action.payload.listName] = action.payload.newList
        },
        update_DELETE_ALL_DATAS:(state,action) => {
            state.datas = {
                userFoldersList:[],
                userItemsList:[],
            }
        },
    },
})

export const UserSliceReducer = UserSlice.reducer
export const {
    update_loadAllDatas,
    update_allDatasLoad,
    update_dataList,
    update_addData,
    update_changeData,
    update_deleteData,
    update_reorderList,
    update_DELETE_ALL_DATAS
} = UserSlice.actions