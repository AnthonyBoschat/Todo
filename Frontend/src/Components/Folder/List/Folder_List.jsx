import React from "react";
import { useSelector } from "react-redux";
import { useDispatch } from "react-redux";
import Button_Folder from "../Button/Folder_Button";
import Creation_Folder from "../Creation/Folder_Creation";
import { update_reorderList } from "../../User/UserSlice";
import { DragDropContext, Droppable } from 'react-beautiful-dnd';
import useFetchRequest from "../../../Utils/useFetchRequest";

export default function Folder_List(){

    const userFoldersList = useSelector(store => store.user.datas.userFoldersList)
    const folderOnCreation = useSelector(store => store.folder.folderOnCreation)
    const dispatch = useDispatch()
    const {fetchRequest} = useFetchRequest()

    const handleOnDragEnd = (result) => {
        const {destination, source} = result
        if(!destination) return
        if(destination.droppableId === source.droppableId && destination.index === source.index)return
        const items = Array.from(userFoldersList)
        const [reorderedItem] = items.splice(result.source.index, 1)
        items.splice(result.destination.index, 0, reorderedItem)
        dispatch(update_reorderList({listName:"userFoldersList", newList:items}))
        fetchRequest("POST", "folder/sort", {newFoldersList:items})
    }


    return(
        <DragDropContext onDragEnd={handleOnDragEnd}>
                <Droppable droppableId="folders">
                    {(provided) => (
                        <div className="listFolder_Display">
                            <ul {...provided.droppableProps} ref={provided.innerRef} className="listFolder_Box">
                
                                    {userFoldersList.length !== 0 && ( // La list de tout les dossiers s'il y en a
                                        userFoldersList.map((folder, index) => (
                                            <Button_Folder index={index} folder={folder} key={folder._id} />
                                        ))
                                    )}
                
                                    {(userFoldersList.length === 0 && !folderOnCreation) && ( // Si aucun dossier ( length === 0), et qu'on est pas en train d'en créer un, indication No folders
                                        <div className="noFolders_Box">
                                            <span>( No folders )</span>
                                        </div>
                                    )}
                                    
                                    {folderOnCreation && ( // Si on créé un dossier, affiche la creation du dossier
                                        <Creation_Folder/>
                                    )}
                                    {provided.placeholder}
                            </ul>
                        </div>
                    )}
                </Droppable>
            </DragDropContext>
        
    )
}