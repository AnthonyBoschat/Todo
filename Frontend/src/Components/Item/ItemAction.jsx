import { useDispatch } from "react-redux";
import { update_addData, update_changeData, update_dataList, update_deleteData } from "../User/UserSlice";
import { update_ItemOnCreation, update_addItemToShow } from "./ItemSlice";
import { useSelector } from "react-redux";

export default function useItem_Action(){

    const userItemsList = useSelector(store => store.user.datas.userItemsList)
    const dispatch = useDispatch()


    const itemAction = {
        create:(data) => {
            dispatch(update_addData({listName:"userItemsList", newData:data}))
            dispatch(update_addItemToShow(data))
            dispatch(update_ItemOnCreation(false))
        },

        delete:(itemID) => {
            const itemIndex = userItemsList.findIndex(item => item._id === itemID)
            dispatch(update_deleteData({listName:"userItemsList", dataIndex:itemIndex}))
        },

        toggleCompleted:(data) => {
            const dataIndex = userItemsList.findIndex(Item => Item._id === data._id)
            dispatch(update_changeData({listName:"userItemsList", dataIndex:dataIndex, newData:data}))
        },

        toggleOnWorking:(data) => {
            data.forEach(data => {
                if(data !== null){
                    const dataIndex = userItemsList.findIndex(Item => Item._id === data._id)
                    dispatch(update_changeData({listName:"userItemsList", dataIndex:dataIndex, newData:data}))
                }
            })
        },

        rename:(data) => {
            const dataIndex = userItemsList.findIndex(Item => Item._id === data._id)
            dispatch(update_changeData({listName:"userItemsList", dataIndex:dataIndex, newData:data}))
        },

        sort:(data) => {
            const objectsDatas = data.reduce((acc, item) => {
                acc[item._id] = item.position
                return acc
            }, {})
            const updatedUserItemsList = userItemsList.map(item => {
                if(objectsDatas[item._id] !== undefined){
                    return {...item, position:objectsDatas[item._id]}
                }
                return item
            })
            dispatch(update_dataList({listName:"userItemsList", newList:updatedUserItemsList}))
            console.log("Requette de réorganisation des tâches effectuer")
        }
    }

    return{
        itemAction
    }
}