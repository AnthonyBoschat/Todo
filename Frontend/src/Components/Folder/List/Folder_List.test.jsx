import { render, screen } from "@testing-library/react"
import { ItemSliceReducer } from "../../Item/ItemSlice"
import Folder_List from "./Folder_List"
import React from "react"
import store from "../../../Redux/store"
import { Provider } from "react-redux"
import { update_folderOnCreation } from "../FolderSlice"
import { update_dataList } from "../../User/UserSlice"

describe("<Folder_List>", () => {
    it("should the noFolder_sign is here when no folder are load in redux", async() => {
        store.dispatch(update_dataList({listName:"userFoldersList", newList:[]}))
         const {container} = render(
            <Provider store={store}>
                <Folder_List/>
            </Provider>
        )
        const noFolder_sign = await screen.getByLabelText("noFolder_sign")
        expect(noFolder_sign).toBeInTheDocument()
        expect(container.firstChild).toMatchInlineSnapshot(`
          <div
            class="listFolder_Display"
          >
            <ul
              class="listFolder_Box"
              data-rbd-droppable-context-id="0"
              data-rbd-droppable-id="folders"
            >
              <div
                aria-label="noFolder_sign"
                class="noFolders_Box"
              >
                <span>
                  ( No folders )
                </span>
              </div>
            </ul>
          </div>
        `)
    })

    it("should folderCreation components are visible when the folderOnCreation are true", async() => {
        
        store.dispatch(update_folderOnCreation(true))
        render(
            <Provider store={store}>
                <Folder_List/>
            </Provider>
        )
        const newFolderCreation = screen.getByLabelText("newFolderCreation")
        expect(newFolderCreation).toBeInTheDocument()
    })

    it("should a list of 3 folder are show when folderDatas is correctly load on Redux", async() => {
        const folderDatas = [
            {_id:"0", name:"Faire les courses"},
            {_id:"1", name:"Faire le m√©nage"},
            {_id:"2", name:"Promener le chien"},
        ]
        store.dispatch(update_dataList({listName:"userFoldersList", newList:folderDatas}))
        store.dispatch(update_folderOnCreation(false))
        render(
            <Provider store={store}>
                <Folder_List/>
            </Provider>
        )
        const folderList = await screen.getByRole("list")
        expect(folderList.children.length).toBe(3)
        const noFolder_sign = screen.queryByLabelText("noFolder_sign")
        expect(noFolder_sign).not.toBeInTheDocument()
    })
})