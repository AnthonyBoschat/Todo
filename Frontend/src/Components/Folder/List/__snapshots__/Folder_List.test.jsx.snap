// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Folder_List> > should a list are visible when redux are feed with correct folders datas 1`] = `"<li data-rbd-draggable-context-id="2" data-rbd-draggable-id="0" tabindex="0" role="button" aria-describedby="rbd-hidden-text-2-hidden-text-4" data-rbd-drag-handle-draggable-id="0" data-rbd-drag-handle-context-id="2" draggable="false" class="folderButton_Box"><div>Faire les courses</div></li><li data-rbd-draggable-context-id="2" data-rbd-draggable-id="1" tabindex="0" role="button" aria-describedby="rbd-hidden-text-2-hidden-text-4" data-rbd-drag-handle-draggable-id="1" data-rbd-drag-handle-context-id="2" draggable="false" class="folderButton_Box"><div>Faire le m√©nage</div></li><li data-rbd-draggable-context-id="2" data-rbd-draggable-id="2" tabindex="0" role="button" aria-describedby="rbd-hidden-text-2-hidden-text-4" data-rbd-drag-handle-draggable-id="2" data-rbd-drag-handle-context-id="2" draggable="false" class="folderButton_Box"><div>Promener le chien</div></li>"`;
